frame/dapp-staking-v3/src/lib.rs

```diff
@@ -0,0 +1,730 @@
+/// Period number type
+pub type PeriodNumber = u32;
+/// Dapp Id type
+pub type DAppId = u16;
+
+/// Distinct period types in dApp staking protocol.
+#[derive(Encode, Decode, MaxEncodedLen, Clone, Copy, Debug, PartialEq, Eq, TypeInfo)]
+pub enum PeriodType {
+    /// Period during which the focus is on voting.
+    /// Inner value is the era in which the voting period ends.
+    VotingPeriod(#[codec(compact)] EraNumber),
+    /// Period during which dApps and stakers earn rewards.
+    /// Inner value is the era in which the Build&Eearn period ends.
+    BuildAndEarnPeriod(#[codec(compact)] EraNumber),
+}
+
+/// Force types to speed up the next era, and even period.
+#[derive(Encode, Decode, MaxEncodedLen, Clone, Copy, Debug, PartialEq, Eq, TypeInfo)]
+pub enum ForcingTypes {
+    /// Force the next era to start.
+    NewEra,
+    /// Force the current period phase to end, and new one to start
+    NewEraAndPeriodPhase,
+}
+
+/// General information & state of the dApp staking protocol.
+#[derive(Encode, Decode, MaxEncodedLen, Clone, Copy, Debug, PartialEq, Eq, TypeInfo)]
+pub struct ProtocolState<BlockNumber: AtLeast32BitUnsigned + MaxEncodedLen> {
+    /// Ongoing era number.
+    #[codec(compact)]
+    pub era: EraNumber,
+    /// Block number at which the next era should start.
+    /// TODO: instead of abusing on-initialize and wasting block-space,
+    /// I believe we should utilize `pallet-scheduler` to schedule the next era. Make an item for this.
+    #[codec(compact)]
+    pub next_era_start: BlockNumber,
+    /// Ongoing period number.
+    #[codec(compact)]
+    pub period: PeriodNumber,
+    /// Ongoing period type and when is it expected to end.
+    pub period_type: PeriodType,
+    /// `true` if pallet is in maintenance mode (disabled), `false` otherwise.
+    /// TODO: provide some configurable barrier to handle this on the runtime level instead? Make an item for this?
+    pub pallet_disabled: bool,
+}
+
+impl<BlockNumber> Default for ProtocolState<BlockNumber>
+where
+    BlockNumber: AtLeast32BitUnsigned + MaxEncodedLen,
+{
+    fn default() -> Self {
+        Self {
+            era: 0,
+            next_era_start: BlockNumber::from(1_u32),
+            period: 0,
+            period_type: PeriodType::VotingPeriod(0),
+            pallet_disabled: false,
+        }
+    }
+}
+
+/// dApp state in which some dApp is in.
+#[derive(Encode, Decode, MaxEncodedLen, Clone, Copy, Debug, PartialEq, Eq, TypeInfo)]
+pub enum DAppState {
+    /// dApp is registered and active.
+    Registered,
+    /// dApp has been unregistered in the contained era
+    Unregistered(#[codec(compact)] EraNumber),
+}
+
+/// General information about dApp.
+#[derive(Encode, Decode, MaxEncodedLen, Clone, Copy, Debug, PartialEq, Eq, TypeInfo)]
+pub struct DAppInfo<AccountId> {
+    /// Owner of the dApp, default reward beneficiary.
+    pub owner: AccountId,
+    /// dApp's unique identifier in dApp staking.
+    #[codec(compact)]
+    pub id: DAppId,
```

Comment:
Forgot to add this one: do we need `id` when they can be identified via `T::SmartContract`?

Reply:
ID was added because of how dApp rewards will be calculated.
Using `T::SmartContract` as identifier works, logically but it takes a lot of space.
To optimize this, a more compact _secundary_ (not sure if that's the right term) is introduced.

Right now it's part of the design (more info in the doc) but it might change in the future.